python/pyspark/sql/pandas/conversion.py

```diff
@@ -375,22 +379,105 @@ def _convert_from_pandas(
         assert isinstance(self, SparkSession)
 
         if timezone is not None:
-            from pyspark.sql.pandas.types import _check_series_convert_timestamps_tz_local
+            from pyspark.sql.pandas.types import (
+                _check_series_convert_timestamps_tz_local,
+                _get_local_timezone,
+            )
             from pandas.core.dtypes.common import is_datetime64tz_dtype, is_timedelta64_dtype
 
             copied = False
             if isinstance(schema, StructType):
-                for field in schema:
-                    # TODO: handle nested timestamps, such as ArrayType(TimestampType())?
-                    if isinstance(field.dataType, TimestampType):
-                        s = _check_series_convert_timestamps_tz_local(pdf[field.name], timezone)
-                        if s is not pdf[field.name]:
-                            if not copied:
-                                # Copy once if the series is modified to prevent the original
-                                # Pandas DataFrame from being updated
-                                pdf = pdf.copy()
-                                copied = True
-                            pdf[field.name] = s
+
+                def _create_converter(data_type: DataType) -> Callable[[pd.Series], pd.Series]:
```

Comment:
Just wondering if you had considered "un-nesting" the Arrow field first, then applying the conversions on flat timestamp fields only, then putting the nested fields back together again? It would be a little more complicated to do this, but would have the benefit of working the same for any level of nested fields, and it's easier to work with Arrow nested fields vs in Pandas.

Reply:
Yes, I had, but I dropped the idea because, in the series of my recent PRs, I wanted to consolidate the conversion logic in various cases, like with/without Arrow, and Spark Connect, or conversions of array, map, and struct types.
Now we can reuse `_create_converter_to_pandas` and `_create_converter_from_pandas` in `pyspark.sql.pandas.types` package in many cases.

The function you pointed was an unfortunate case, I couldn't consolidate only that part. We might need to revisit here.

Reply:
No problem, that makes sense